validator/src/main/java/com/amazon/aoc/validators/TraceValidator.java

```diff
@@ -113,30 +134,33 @@ public void validate() throws Exception {
   }
 
   private Map<String, Object> flattenDocument(List<Segment> segmentList) {
-    // have to sort the segments by start_time because
-    // 1. we can not get span id from xraysdk today,
-    // 2. the segments come out with different order everytime
-    segmentList.sort(
-        (segment1, segment2) -> {
-          try {
-            Map<String, Object> map1 =
-                new ObjectMapper().readValue(segment1.getDocument(), Map.class);
-            Map<String, Object> map2 =
-                new ObjectMapper().readValue(segment2.getDocument(), Map.class);
-            return Double.valueOf(map1.get("start_time").toString())
-              .compareTo(Double.valueOf(map2.get("start_time").toString()));
-          } catch (Exception ex) {
-            log.error(ex);
-            return 0;
-          }
-        });
+    List<Entity> entityList = new ArrayList<>();
 
-    // build the segment's document as a jsonarray and flatten it for easy comparison
-    StringBuilder segmentsJson = new StringBuilder("[");
+    // Parse retrieved segment documents into a barebones Entity POJO
     for (Segment segment : segmentList) {
-      segmentsJson.append(segment.getDocument());
-      segmentsJson.append(",");
+      Entity entity;
+      try {
+        entity = MAPPER.readValue(segment.getDocument(), Entity.class);
```

Comment:
Also, not sure how often this scenario would occur but do you think mapper should not be able to map some values from document to the class you have created? If this is the case then we will also have to do some handling on fields because for example let's say if mapper didn't translate `name` then validation would fail if we are checking `name` value in stored validation template.

Reply:
Hmm well `Entity` has all the fields that we support in the (sub)segment model, in addition to `inferred` which is added by the backend service. If we want to validate against any field that isn't yet a part of the Entity class, we can just add it to the Entity class so it's read properly then validate it normally. Like I mentioned new fields for entities are allow-listed pretty rarely, so I think it's acceptable to have to update that class when they do come in.

Reply:
sounds fair.